### PR TITLE
Package Manager MVP

### LINKED ISSUE #353: ability to add a package to build based on path to build.zig
This can be implemented in the zig build system:

Add a package by giving the path to a directory which has build.zig in it, an alias name which will be used as the `@import(...)` string, a set of build options (the same options available from the command line, visible from the `--help` menu), and an expected hash.

The build system does these steps when building:
 * Copies the entire directory to `./zig-pkg/alias/` inside the current project's build directory. If the directory already exists and contains the correct expected hash, this copy operation can be skipped. If the copy is performed, compute the hash while doing the copy, and ensure that the expected hash is correct. If not, discard the data and produce an error. If the existing directory contains a different hash than the expected one, remove it and proceed as if there were no directory there.
 * builds `./zig-pkg/alias/build.zig` but instead of with build_runner.zig it uses its own runner, which does the following:
   * runs the `pkg` step. note that this recursively invokes this whole process. If the `pkg` step does not exist, it means no build step is necessary, and the copied source files is sufficient.
   * identifies the package's root source file and reports it via IPC
   * reports the package's dependencies via IPC
 * using all the collected information, correctly reports the `--pkg-begin` and `--pkg-end` arguments to the zig compiler

You can see how this fits into depending on URLs as packages. Instead of copying a directory, it would download and unpack a file, or clone/checkout a repository, rather than copy files, and everything else is the same. We can save that for another issue.

Bonus feature: when a hash is not matched based on a directory copy or a download, save the data in a LRU cache in case the user decides to update the expected hash to this value, and check this LRU cache before re-downloading/copying the data redundantly.

### LINKED ISSUE #368: Import a build.zig file as a build step
Kinda like this:
https://github.com/zig-lang/zig/blob/master/test/build_examples.zig

But included in the standard build system.

### LINKED ISSUE #943: package manager
[Latest Proposal](https://github.com/ziglang/zig/issues/943#issuecomment-386458021)

----

Zig needs to make it so that people can effortlessly and confidently depend on each other's code.

~Depends on #89~

### PR DESCRIPTION
**Pull request by andrewrk**
### Demo

```
andy@ark ~> cd tmp/
andy@ark ~/tmp> git clone https://github.com/andrewrk/ffmpeg --depth 1
Cloning into 'ffmpeg'...
remote: Enumerating objects: 4138, done.
remote: Counting objects: 100% (4138/4138), done.
remote: Compressing objects: 100% (3823/3823), done.
remote: Total 4138 (delta 864), reused 1687 (delta 302), pack-reused 0
Receiving objects: 100% (4138/4138), 13.30 MiB | 10.58 MiB/s, done.
Resolving deltas: 100% (864/864), done.
andy@ark ~/tmp> cd ffmpeg
andy@ark ~/t/ffmpeg (main)> time zig build

________________________________________________________
Executed in   35.84 secs    fish           external
   usr time  409.13 secs  994.00 micros  409.13 secs
   sys time   35.68 secs  977.00 micros   35.68 secs

andy@ark ~/t/ffmpeg (main)> find zig-out/
zig-out/
zig-out/lib
zig-out/lib/libffmpeg.a
andy@ark ~/t/ffmpeg (main)> cat build.zig.ini
[package]
name=libffmpeg
version=5.1.2

[dependency]
name=libz
url=https://github.com/andrewrk/libz/archive/f0e53cc2391741034b144a2c2076ed8a9937b29b.tar.gz
hash=c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae

[dependency]
name=libmp3lame
url=https://github.com/andrewrk/libmp3lame/archive/497568e670bfeb14ab6ef47fb6459a2251358e43.tar.gz
hash=9ba4f49895b174a3f918d489238acbc146bd393575062b2e3be33488b688e36f
andy@ark ~/t/ffmpeg (main)> cat build.zig | head -n36
const std = @import("std");

pub fn build(b: *std.build.Builder) void {
    const target = b.standardTargetOptions(.{});
    const mode = b.standardReleaseOptions();

    const libz_dep = b.dependency("libz", .{});
    const libmp3lame_dep = b.dependency("libmp3lame", .{});

    const lib = b.addStaticLibrary("ffmpeg", null);
    lib.setTarget(target);
    lib.setBuildMode(mode);
    lib.linkLibrary(libz_dep.artifact("z"));
    lib.linkLibrary(libmp3lame_dep.artifact("mp3lame"));
    lib.linkLibC();
    lib.addIncludePath(".");
    lib.addCSourceFiles(&avcodec_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_avcodec",
    });
    lib.addCSourceFiles(&avutil_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_avutil",
    });
    lib.addCSourceFiles(&avformat_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_avformat",
    });
    lib.addCSourceFiles(&avfilter_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_avfilter",
    });
    lib.addCSourceFiles(&swresample_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_swresample",
    });
    lib.addCSourceFiles(&swscale_sources, ffmpeg_cflags ++ [_][]const u8{
        "-DBUILDING_swscale",
    });
    lib.install();
}
andy@ark ~/t/ffmpeg (main)> time zig build

________________________________________________________
Executed in  381.90 millis    fish           external
   usr time    0.47 secs    921.00 micros    0.47 secs
   sys time    1.09 secs      0.00 micros    1.09 secs
```

You can play with this yourself:
 * https://github.com/andrewrk/ffmpeg
 * https://github.com/andrewrk/libz
 * https://github.com/andrewrk/libmp3lame

Keep in mind, some enhancements to TLS need to be made before downloading https will work for non-Linux users:
 *  #14169
 *  #14170 (there's an open PR, I will merge it soon)

### Why .ini format for the declarative file?

Here are options that I considered and reasons for why I ultimately rejected them:

 * JSON
   - unfortunate that trailing commas are not allowed, and comments are not allowed.
     - json5 exists but a lot of tooling does not expect it. and what's the extension? .json5 or .json? either one is a bit problematic.
   - what's javascript object notation doing in my zig codebase? I don't have any javascript objects to describe.
   - numbers limited to 53 bits sometimes? there is some janky stuff with json.
 * TOML
   - despite the O standing for "obvious", the format is over-complicated. There are multiple ways to convey the same meaning which leads to two problems:
     - it can be confusing when something in a different location in the file affects something non-local.
     - when a machine wants to mutate the data and write it back out to disk, there is ambiguity about how to render the values.
   - the format has unnecessary stuff that we don't need, making tooling more cumbersome.
 * CSV
   - really the only data we need is tables of stuff. CSV is surprisingly a good fit. except that there are multiple tables, not just one
   - CSV is not really nice to edit with a text editor. silly to require a GUI to edit this data
 * YAML
   - nothing but a labyrinth of white space and inscrutable symbols, a prison for the unwary coder. It mocks my intelligence with its simplistic structure and minimal syntax, yet it still manages to ensnare me in its deadly embrace. I curse the day I first laid eyes on that ridiculous format, and the false promises of ease and clarity it offered me. It has only brought me pain and frustration, with its inconsistent indentation and subtle errors lurking in every line.
 
So here's what I came up with: we will specify a very restricted subset of .ini. The restricted subset of .ini that zig supports for this file is trivial to parse. It's also technically a subset of TOML, and intended to be a subset of most .ini parsers in the wild, so existing libraries & tools can be reused, probably.

The filename, `build.zig.ini` is intended to imply that it is an appendage to `build.zig` because that is exactly what it is. The real, actual file that signifies a zig package is a build.zig, and the existence of this extra file is bonus - it is for the case of declarative information that we want to expose without requiring execution of zig code.

In fact, if you look at the example above - libz and libmp3lame - these packages do not actually have a `build.zig.ini`. They do not have any declarative information to declare. This is perfectly legal, and has the same repercussions as if you copy-pasted the files directly into VCS.

It's very easy to bikeshed this topic (the topic of chosen file format). Feel free to discuss, but please make sure new comments add something to the discussion, not repeat something already said.

### Follow-up Issues

 * [x] #14278
 * [x] #14279
 * [x] #14280
 * [x] #14281
 * [x] #14282
 * [x] #14283
 * [x] #14284
 * [x] #14285
 * [ ] #14286
 * [ ] #14287
 * [ ] #14288
 * [ ] #14289
 * [x] #14290
 * [ ] ~competing proposal: use s-expressions and the file can be named `build.zig.sex`~ nah
 * [ ] #14291
 * [ ] #14292
 * [ ] #14293
 * [ ] #14294
 * [ ] #14295
 * [ ] #14296
 * [ ] #14297
 * [x] #14298
 * [ ] sigstore support. I will leave it to @jedisct1 to write up a proposal for this.
 * [x] #14299
 * [x] #14300
 * [ ] #14307
 * [x] #14308
 * [ ] #14309
 * [x] #14310
 * [x] #14311
 * [ ] #14312
 * [ ] #14314

----

closes #353
closes #368
closes #943

### CODE REVIEW

**Comment by motiejus** [comment]
Congratulations with the start!

Future proposal: would you consider an optional vendoring of dependencies in an adjacent git ref ([background](https://jakstys.lt/2022/smart-bundling/)) ? In this example:

```
# build.zig.ini
[dependency]
name=libz
git_url=https://github.com/andrewrk/libz
git_sha=f0e53cc2391741034b144a2c2076ed8a9937b29b
vendor=git-ref

[dependency]
name=libmp3lame
git_url=https://github.com/andrewrk/libmp3lame
git_sha=497568e670bfeb14ab6ef47fb6459a2251358e43
vendor=git-ref
```

When zig finds a "git-ref"-vendored dependency, instead of downloading from github, it checks it out from the project's git tree, similarly how [`git-subtrac`](https://apenwarr.ca/log/20191109) does things, but without submodules. You rejected a similar idea in person in April 2022, because it required a user to run a git command to check it out. In this case, `zig` could do it itself, which makes a difference.

Main benefits:
- dependencies never disappear, as they come together with the project.
- works offline: one can clone the repository, disconnect from the Internet, do `zig build`, and it will work.

I would be willing to work on this assuming there is no outright rejection. Let me know!

**Comment by andrewrk** [comment]
That looks like a neat approach. If I understand correctly, that approach can already be done even with this minimal MVP. In fact it's exactly what is being done in my example above.

Perhaps what is missing would be achieving this, but for a URL that only supports the git protocol and does not have that convenient https download that GitHub has made available.

Or perhaps what is missing, more importantly, would be some kind of detection that the URL is in fact a git repository that is already cloned and corresponds to the same one that the `zig build` command is being executed in. The system could then avoid the internet entirely to fetch the dependency.

Let me know what you think.

**Comment by lukehinds** [comment]
I get this is a PoC / prototype, but please don't leave crypto signing / verification as an afterthought when you ship something (just as rust, python, npm (in fact all of them) have done), as it's always tricky to retrofit. 

happy to work with you from the sigstore community

**Comment by andrewrk** [comment]
The current plan, as already implemented in this branch, is Trust On First Use. If you would like to make a case for doing something more complicated than this, you are welcome to send some learning materials our way.

**Comment by lukehinds** [comment]
> The current plan, as already implemented in this branch, is Trust On First Use. If you would like to make a case for doing something more complicated than this, you are welcome to send some learning materials our way.

How's that work @andrewrk , typically ToFu means you cache an identifier for subsequent use (like ssh, I select yes and tofu is set up for later connections)? 

I will look to gather some materials for you.

**Comment by motiejus** [comment]
> That looks like a neat approach. If I understand correctly, that approach can already be done even with this minimal MVP. In fact it's exactly what is being done in my example above.

I think we misunderstood each other. The repository would look like this (feel free to poke at it):

```
$ git clone github.com/motiejus/ffmpeg
$ $ git ls-tree f0e53cc2391741034b144a2c2076ed8a9937b29b | head -5
100644 blob 93c1b5f431849d1e64c92cad8ad04a8ca1f4b5e6    .gitignore
100644 blob ab8ee6f71428c3e4646b12b64ad387cde33462e5    LICENSE
100644 blob e9f5b2298d8948f46b2803070e26b29cb41fd0f2    README
100644 blob d0be4380a39c9c5bf439b1552c43585b5aafad0a    adler32.c
100644 blob 30c19fc64410572400f6bb0e89a2c6e7adf3bd13    build.zig
```

Note that f0e53cc2391741034b144a2c2076ed8a9937b29b comes from libz. And the `main` of my repository is unchanged; so does f0e53cc2391741034b144a2c2076ed8a9937b29b - it points to the original code.

> Perhaps what is missing would be achieving this, but for a URL that only supports the git protocol and does not have that convenient https download that GitHub has made available.

If zig recognizes the repository as "git-vendored", zig should, instead of downloading it, take the git tree from the same repository.
 
However, there are two caveats:

1. Git will not download the dependencies with `--depth=1`. 
2. The developer, after changing the dependencies, should push `main.trac` as well. This can be trivially detected and "fixed-up" with an internet connection.

**Comment by nicoburns** [comment]
If you'd be interested in a new config format to adopt, I've been very impressed with the design of the "gura" format (https://github.com/gura-conf/gura). It seems to combine a lot of the best of JSON, YAML and TOML while being relatively simple. I don't think it's seen widespread adoption anywhere though, so tooling support may be limited.

**Comment by mrakh** [comment]
A couple of concerns that I have:

- Ensuring first-class support for non-git repositories. SVN and Mercurial come to mind.
- Relying on arbitrary code execution to resolve dependencies. I consider this to be a Very Bad Idea<sup>TM</sup>, as it makes static analysis intractable, and provides an easy vector for malicious packages to compromise users. In my opinion, required dependencies should be statically declared in the ini file.

**Comment by erikarvstedt** [comment]
Using this with package managers like Nix or Guix requires support for separating the build into two steps:
1. An online dependency fetching step, whose output is solely determined by the content hashes.
2. An offline build step, which receives step 1. as an input.

Step 1. could be implemented by zig (like `cargo vendor`), but it doesn't have to. At least, there should be a specification of the format in which the deps are passed to step 2.

**Comment by nhh** [comment]
Zig dependency file naming proposal: Ziggyfile

**Comment by uranusjr** [comment]
I don’t think the file format is valid ini. All the ini parsers I know of either outright reject the duplicated `[dependency]` sections, or merge all of them into one section (i.e. you end up with only the last dependency section parsed). If you are to stick with the one-section-per-dependency syntax, TOML is the only viable choice I know of.

Some other thoughts coming from another packaging ecosystem (I maintain pip and contributes to many specs and tools in Python packaging): If you are to support installing from VCS, strictly only support commit IDs (or equivalent), at least for now. Anything that can potentially change (especially branches, but also others) would introduce endless cachability and reproducibility issues. You could leave room and figure them out later, but don’t commit (not a pun) to supporting them too early. Also big +1 on not relying on arbitrary code execution, it’s definitely one of the worst things you could do.

Another thing worth leaving room for is to allow multiple sources to satisfy one dependency (that the resolver can choose on build-time based on various criteria, such as OS, CPU, GPU, etc.). Eventually you’ll reach a point where people wants to distribute pre-built binaries, and it’ll cause much growing pain to change a one-to-one assumption to many-to-one unless you plan ahead.

**Comment by mlugg** [comment]
I'm interested in how the args are planned to work. From what I'm reading, will it be command-line-style arguments, and they will all be applied to the child builder just as if those arguments were passed? I think that'll work okay for my use cases, even if the usage seems a bit odd in my head. Is there really any need to pass options other than `-Dfoo=bar` to dependencies? If not, maybe the args structure could be special-cased as just a big key-value list (`[]const struct{ []const u8, []const u8 }`)?

File: lib/std/tar.zig [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    pub fn fileSize(header: Header) !u64 {
+        const raw = header.bytes[124..][0..12];
+        const ltrimmed = std.mem.trimLeft(u8, raw, "0");
+        const rtrimmed = std.mem.trimRight(u8, ltrimmed, "\x00");
+        if (rtrimmed.len == 0) return 0;
+        return std.fmt.parseInt(u64, rtrimmed, 8);
+    }
+
+    pub fn is_ustar(header: Header) bool {
```
Comment (truemedian) [inline lib/std/tar.zig:31]: This doesn't follow the style guide.

**Comment by losvedir** [comment]
So exciting to see progress towards a package manager! (For context, here is my [only comment](https://github.com/ziglang/zig/issues/943#issuecomment-878305090) on the other PR, where I advocate for a package manager in the style of Elixir.)

I'm curious how this will extend to transitive dependencies. The example here shows two dependencies that don't themselves have any dependencies (which is itself kind of neat that the deps don't need to be a part of this system). But I'm wondering how it will work when the library you want to use depends on other libraries. I can think of three general approaches:

1. Don't do that. Every project is either an app that uses libraries, or a library that has no dependencies. As seen here.
2. Libraries that each use the same transitive library will have their own copies, possibly on different versions. This is how `npm`  does it, and I think part of what leads to huge amounts of churn in that ecosystem, a gigantic lock file, and a `node_modules` folder that quickly explodes in size.
3. Transitive dependencies are shared, so your application's libraries need to coordinate their requirements, and install a single compatible version that works for everyone (or raises an issue if that can't be done). This is how Elixir does it, which leads to a small lock file, a low amount of dependency code, but some human effort to ensure that dependencies share compatible transitive dependencies and are upgraded in sync. But this requires a way to specify dependency requirements, rather than links to specific bits of code, and some sort of resolution system.

**Comment by jedisct1** [comment]
@lukehinds It's a new package manager, and an opportunity to get things right from the beginning.

And yep, Sigstore is something we should absolutely support. Happy to work with you on that. Maybe we can even reuse some of the work we did on wasm signatures (multiple signers, policies, etc. would apply very well, and would be a significant improvement over other package managers).

**Comment by jayschwa** [comment]
> Another thing worth leaving room for is to allow multiple sources to satisfy one dependency (that the resolver can choose on build-time based on various criteria, such as OS, CPU, GPU, etc.).

Pre-built artifacts aside, multiple URIs for source archives is also forward-looking. Example:

```
# Slow upstream server in basement that will disappear in 10 years.
ftp://upstream-hermit.dev/project/source.tar.gz

# Popular, fast hosting site with fickle content moderation.
https://github.com/project-mirror/archive.tar.gz

# Another popular, fast hosting site.
https://gitlab.com/project-mirror/whatever.tar.gz

# Martian colonists can fetch from their local swarm.
magnet:contentaddress
```

Multiple HTTP mirrors could be useful now and other protocols can be ignored. Down the line, Zig could integrate with operating systems' registered protocol handlers or allow users to configure an arbitrary external command for a protocol.

My other suggestion is to specify hashing algorithm(s) in-band (e.g. replace `hash` key with `sha256sum` or whatever). Git's glacial (stalled?) migration from SHA1 to SHA256 is a relevant cautionary tale.

**Comment by deflock** [comment]
```ini
[dep:libz]
url=https://github.com/andrewrk/libz/archive/f0e53cc2391741034b144a2c2076ed8a9937b29b.tar.gz
sha1=c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae
```
🤔

**Comment by andrewrk** [comment]
Thank you, all, for your comments. I have digested them into an actionable set of follow-up issues, to be filed upon merging this branch (scroll up to see my edit).

**Comment by asdf8dfafjk** [comment]
Hjson? Tooling shouldn't decide. Use what's best

**Comment by lukehinds** [comment]
As promised some material on the signing / verification:

* The NPM RFC (merged, will be implemented) https://github.com/npm/rfcs/pull/626
* The work for integration to Python https://blog.trailofbits.com/2022/11/08/sigstore-code-signing-verification-software-supply-chain/
* Pre-RFC for rust that just got underway https://internals.rust-lang.org/t/pre-rfc-using-sigstore-for-signing-and-verifying-crates/18115

A quick 'what is sigstore'

It's an OSS project with open governance and funded by the [OpenSSF](https://openssf.org/). You can think of sigstore, as the let's encrypt for software signing. The links above give a good technical overview.

**Comment by nektro** [comment]
@luickk `gist.github.com` pastes are all git repos under the hood and would be supported once git dependencies are supported. for some prior art [zigmod](https://github.com/nektro/zigmod) (my 3rd party package manager) already supports this, and I even use a gist as a dependency in zigmod itself. see https://github.com/nektro/zigmod/blob/master/zig.mod

**Comment by MoAlyousef** [comment]
If git dependencies get supported, I think it would be nice if the package manager also automatically fetches the submodules recursively.

**Comment by motiejus** [comment]
I believe this is related to my points above?

>  ability to fetch git commits from the clone that contains build.zig without network access

Having spent some time discussing here and on IRC, I suggest the following enhancement:

    If a dependency's GIT_HASH can be found in the project's git tree, use that.

This will allow:
- have have *all* project dependencies in the same repository
- non-opinionated ways to store dependencies in the git tree (single branch, multiple branches, tags, different remotes, etc).
- the enhancement itself *feels* small and unopionated. I may be missing something of course.

I would be willing to implement this myself.

@andrewrk let me try to dissect your latest comment on IRC about this:

> to be honest, this situation seems to me to be rare,

Thank you for your honesty! It is rare, but not original. Here are some similar projects:
- https://github.com/tailscale/go-mod-archiver
- https://github.com/apenwarr/git-subtrac

Note: with the above enhancement subtrac-enabled projects would *just work* out of the box.

> straightforward to resolve,

I don't see a way to resolve this *conveniently to the end-user* without building this into the package manager. Unless package manager supports "plugins" or so, which is far fetched. :)

My end goal is:
```
$ git clone https://project.git; cd project
# go offline
$ zig build
```

> and inconsequential

I guess this depends on the environment in which we operate. For me it *is* important to protect myself from dependencies going away (e.g. removed git repository) without setting up extra infrastructure just for that: e.g. caching proxies or manually mirroring everything.

**Comment by cryptocode** [comment]
For config, I really like the _s-expr_ idea, but also think a TOML _subset_ would work nicely (not a full parser, which is rejected for good reasons)

Below is a valid TOML version of the example, and no other syntax would be needed. The parser should be nearly as simple as the ini version.

```toml
[package]
name="libffmpeg"
version="5.1.2"

[[dependency]]
name="libz"
url="https://github.com/andrewrk/libz/archive/f0e53cc2391741034b144a2c2076ed8a9937b29b.tar.gz"
hash="c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae"

[[dependency]]
name="libmp3lame"
url="https://github.com/andrewrk/libmp3lame/archive/497568e670bfeb14ab6ef47fb6459a2251358e43.tar.gz"
hash="9ba4f49895b174a3f918d489238acbc146bd393575062b2e3be33488b688e36f"
```

The format of `build.zig.toml` would be a proper subset of TOML, with just enough to support Zig's needs just like the ini format.

Some benefits IMO:
* Any TOML parser accepts this, and all major editors/github have highlighting support
* The `[[dependency]]` header syntax drives home the fact that this is potentially a list, so it can be repeated
* It's obvious that values are just strings

**Comment by andrewrk** [comment]
@motiejus I think our viewpoints are more aligned than it seemed at first. To be clear, my comments about the situation that I thought was rare, straightforward to resolve, and inconsequential was in response to your concern about someone forgetting to push a branch that contained the dependency; *not* about the use case itself.

You have worded the enhancement like this

> If a dependency's GIT_HASH can be found in the project's git tree, use that.

I worded it like this: 

>  ability to fetch git commits from the clone that contains build.zig without network access

Let us try to tease out the differences in how we are thinking about this problem, because I think we have the same goals in mind.

I'll start: The `hash` field, as it stands, is independent from the method of fetching. This allows mirrors to work, even when mirrors have different protocols than each other, and even when, for example, git protocol fetches extra unwanted files that are not part of the package. Excluded files are deleted from the file system prior to hashing. The hash field is the actual identifier that is used to determine if the cache has already downloaded a particular dependency.

So, looking at your original code snippet, we need to keep the original hash:

```diff
[dependency]
name=libz
git_url=https://github.com/andrewrk/libz
git_sha=f0e53cc2391741034b144a2c2076ed8a9937b29b
vendor=git-ref
+hash=c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae
```

Cleaning up the url a bit:

```diff
[dependency]
name=libz
-git_url=https://github.com/andrewrk/libz
-git_sha=f0e53cc2391741034b144a2c2076ed8a9937b29b
-vendor=git-ref
+url=https+git+ref://github.com/andrewrk/libz#f0e53cc2391741034b144a2c2076ed8a9937b29b
hash=c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae
```

The other problem to be solved is that we still want transitive dependencies to be fetched from the root level VCS, without requiring an application-specific patch to the leaf dependency's manifest file. This implies that more data is needed in the application-level manifest to override hashes and where to find them. Imagine an application that does not directly depend on libz, but libz is a transitive dependency. The application's manifest might include something like this:

```
[mirror]
hash=c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae
url=https+git+ref://github.com/motiejus/magnum-opus#f0e53cc2391741034b144a2c2076ed8a9937b29b
```

This would provide an additional mirror for any package which has a corresponding hash of `c9b30cffc40999d2c078ff350cbcee642970a224fe123c756d0892f876cf1aae`. The url points into the top-level application's git URL rather than at the dependency's VCS URL. The URL can be matched against what is found in `.git/config` to determine if it can be found locally. The `https+git+ref` scheme tells Zig to use the git-over-https protocol, and then further to treat a particular git ref as the directory of files that is the package. Here I take advantage of the fragment portion of the URI for the commit sha.

This additional mirror would have higher priority than the others for two reasons:
 * Explicit mirrors like this would have higher priority
 * Local mirrors are prioritized over network fetches

One more important observation: with this system, a package which adds extra mirrors such as this would *still be viable dependency itself*. Other solutions to this problem, such as referencing "the git clone containing this package" would not be viable since they themselves could not be then stored in a different git clone and then used as a dependency. Also, it's nice for packages not to assume that they are stored in any particular VCS.

**Comment by yvz5** [comment]
Isnt it against the idea of using zig to build ? Is there a reason why we would actually need an extra file ? We might as well just use build.zig

**Comment by andrewrk** [comment]
There is a lot of value to be gained in declarative information for certain things. In particular, it's much faster to fetch all transitive dependencies if you can learn what network URLs to go download, or the file system paths to check the cache for, without running a compiler.

Furthermore, third party tools such as package indexes want to understand information about packages without executing untrusted code. Even first-party tools such as zig itself will use this declarative information to provide package management tooling.

The manifest file describes the set of *potential* dependencies, and the build.zig logic can still pick and choose which ones to use depending on build configuration.

**Comment by likern** [comment]
> There is a lot of value to be gained in declarative information for certain things. In particular, it's much faster to fetch all transitive dependencies if you can learn what network URLs to go download, or the file system paths to check the cache for, without running a compiler.
> 
> Furthermore, third party tools such as package indexes want to understand information about packages without executing untrusted code. Even first-party tools such as zig itself will use this declarative information to provide package management tooling.
> 
> The manifest file describes the set of _potential_ dependencies, and the build.zig logic can still pick and choose which ones to use depending on build configuration.

But we can use subset of Zig syntax. Let's say comptime array of structs of special form.

```
const { Dependency } = @import("std"). package;

pub const deps = [_]Dependency{
    Dependency {.name = "utils", .hash = ..., .url = ...},
    Dependency {.name = "coolthing", .hash = ..., .url = ...},
    ...
};
```
No need to learn something new, directly zero-learning path. Since we strive to declarative approach,  this already will be very simple. It is understandable as .ini file (especially for Zig developers).

The syntax is pretty neat, if it's not, we should improve Zig ;))

Since we already will implement parser for config, we can just write special parser for very strict subset of Zig lang (utilizing all engineering for full-blown Zig).

We can gradually relaxing rules, if needed. Parser and third-party tools could easily parse and understand this file (if no string interpolations are allowed).

But we can (preferable) provide just official zigpkg package manager, which can output json just by `zigpkg --json` or toml, or whatever we want. We can even provide .so lib so that third-party tools could use as a library.

And this file still stays declarative.

**Comment by EleanorNB** [comment]
I would like to submit my vote against subsets/supersets of languages for manifests. We don't want external tools to have to understand and implement ".ini with Zig extensions" or ".toml with Zig retractions". It's messy, it's a headache for them, and it'll be a headache for us too once people start complaining that their preexisting INI parsers or TOML generators don't work.

If we can't find a preexisting format that fits our needs and values without modification, my vote is we make our own s-expression-based format, since it'll be simplest to make tools for.

**Comment by ecstrema** [comment]
my two cents: if the only problem with JSON is that it doesn't support comments and trailing commas, and we don't want json5 or anything because it's non-standard, then creating a new specification for a kind-of-ini file is not what I would call the obvious path.

**Comment by EleanorNB** [comment]
Also, please let's separate fetching entirely from building. I'd prefer to know when my tools are going to be accessing the network, so if I try to build and I don't already have the dependencies I'd rather get an error than a surprise long build. Explicit over implicit and such. (Also that'll be better for external dependency management systems as previously mentioned.)

**Comment by ecstrema** [comment]
Sorry if this doesn't sound serious, but what about zson? zig's object notation. Basically json but with commas and comments. 

JSON still is, in my opinion, the most standard and clearest way to describe objects.

**Comment by EleanorNB** [comment]
@ecstrema I don't think that would work very well. JavaScript objects are schemaless as JavaScript is dynamically typed; Zig tuples are not as Zig is not. How would we represent such concepts as "array of dependencies" in-band? For that matter, how would we distinguish such concepts as "array" from "struct"?

**Comment by andrewrk** [comment]
> Sorry if this doesn't sound serious

[lulz](https://github.com/ziglang/zig/issues/14290)

**Comment by kuon** [comment]
I do not know if I should create a proposal for that, but there is one thing I think could be nice is to have sponsoring directly supported in the package manager.

For example, if you run `mix hex.sponsor` in elixir, it will print a list of all packages your project depends on that accept sponsoring. JS has `npm fund`.

I don't really know exactly what form it should/would take, but if it can help funding open source I think it is a nice addition.

**Comment by EleanorNB** [comment]
@kuon People who care about such a thing will be comfortable using an external tool for that (which they'll have to do anyway to actually sponsor anyone). We could provide such a tool, but not in the package manager. It is not at all the job of a package manager to do that.

**Comment by uranusjr** [comment]
Again coming from another packaging ecosystem, I have to also put my vote against using code for metadata configuration, even a subset of it. Most of the hurdle users have to effective write a package metadata file is not the language, but what heys exactly are needed for what, and most people simply don’t do packaging enough to remember what’s needed. It is therefore not particularly useful to optimise how easy users can memorise and write the declaration—they can’t, no matter what you use. It would be better to optimise for

1. Machine writability (so people can use interactive tools to generate the file)
2. Human readability (so people can easily inspect the file, understand whether it needs modification, and perform simple edits by hand when appropriate)
3. Machine readability (so the interactive tools can _modify_ the file; this is less important since non-trivial modifications are uncommon)

The originally proposed section-key-value non-ini format actually satisfies these quite well. The main problem of it would be tooling and extendability in the future, but as long as you spec it well enough (big if!) this is at least solvable.

**Comment by andrewrk** [comment]
For those who want to influence Zig's package manager as the project moves forward, here is how to do it:

 * familiarize yourself with all the follow-up issues that I filed (links above).
 * come up with a concrete, actionable proposal that has clear criteria for being met. Use examples.
 * comment on an existing issue if applicable, or file a new one otherwise. Explain clearly and patiently, with the goal of communicating how things should be different. If I don't really understand your suggestion then I will ignore it.

**Comment by motiejus** [comment]
> @motiejus I think our viewpoints are more aligned than it seemed at first. To be clear, my comments about the situation that I thought was rare, straightforward to resolve, and inconsequential was in response to your concern about someone forgetting to push a branch that contained the dependency; _not_ about the use case itself.

Now I can  see why my suggested enhancement is a subset of your proposal. Thank you for your thoughtful reply!

> One more important observation: with this system, a package which adds extra mirrors such as this would _still be viable dependency itself_. Other solutions to this problem, such as referencing "the git clone containing this package" would not be viable since they themselves could not be then stored in a different git clone and then used as a dependency. Also, it's nice for packages not to assume that they are stored in any particular VCS.

This is an excellent point, and a way to resolve it, which I did not originally consider.

**Comment by andrewrk** [comment]
I updated https://github.com/ziglang/zig/projects/4 to help organize efforts related to package management.

**Comment by postspectacular** [comment]
I couldn't find a reference in any of the earlier comments/issues and didn't want (yet) to create a new issue, but is there any planned support for monorepos, i.e. git repos hosting multiple projects/packages which are being developed in parallel and which _might_ have some inter-dependencies with other packages in that repo? An extreme case is [Google's monorepo](https://medium.com/qe-unit/how-google-does-monorepo-revisited-8c793be20344), but I'm interested in growing [my own super-early stage Zig monorepo](https://github.com/thi-ng/zig-thing), which currently is using [gyro](https://github.com/mattnite/gyro), but I'm also considering adding more hybrid Zig/TypeScript libs (for WASM apps) to my main monorepo (see below)...

What I think is needed for this general use case (and couldn't find info on): When specifying a dependency on a package from such a monorepo, we'd also be required to provide a subdirectory or direct path to that package's own `build.zig.ini` file... has this already been considered? If not, what would you require to push this forward?

Ps. I perfectly understand there's no major interest in these setups, but it'd be nice if they aren't categorically excluded from the outset (and I'm not saying that they are!) 😉 FWIW I've been using a monorepo for the past 5 years for [all my TypeScript work](https://github.com/thi-ng/umbrella) and wouldn't go back. It's IMHO the only _feasible/sensible_ way to manage & automate scaffolding, maintenance, version bumps, change logs, CI/releases of all the 175+ packages in there (rather than requiring the same number of individual git repos and duplicating a ton of infrastructure setup)...

**Pull request merged**